app/questionnaire/questionnaire_schema.py

```diff
@@ -839,17 +883,113 @@ def _populate_when_rules_section_dependencies(self) -> None:
             when_rules = self.get_values_for_key(section, "when")
             rules: list = list(when_rules)
 
-            if rules_section_dependencies := self._get_rules_section_dependencies(
-                section["id"], rules
-            ):
+            (
+                rules_section_dependencies,
+                rule_section_dependencies_for_progress_value_source,
+                rule_block_dependencies_for_progress_value_source,
+            ) = self._get_rules_section_dependencies(section["id"], rules)
+
+            if rules_section_dependencies:
                 self._when_rules_section_dependencies_by_section[
                     section["id"]
                 ] = rules_section_dependencies
 
+            if rule_section_dependencies_for_progress_value_source:
+                self._when_rules_section_dependencies_by_section_for_progress_value_source |= (
+                    rule_section_dependencies_for_progress_value_source
+                )
+
+            if rule_block_dependencies_for_progress_value_source:
+                self._populate_block_dependencies_for_progress_value_source(
+                    rule_block_dependencies_for_progress_value_source
+                )
+
+    def _populate_block_dependencies_for_progress_value_source(
+        self,
+        rule_block_dependencies_for_progress_value_source: dict[
+            str, dict[str, set[str]]
+        ],
+    ) -> None:
+        for dependent_section in rule_block_dependencies_for_progress_value_source:
+            if (
+                dependent_section
+                in self._when_rules_block_dependencies_by_section_for_progress_value_source
+            ):
+                for block_id in rule_block_dependencies_for_progress_value_source[
+                    dependent_section
+                ]:
+                    if (
+                        block_id
+                        in self._when_rules_block_dependencies_by_section_for_progress_value_source[
+                            dependent_section
+                        ]
+                    ):
+                        self._when_rules_block_dependencies_by_section_for_progress_value_source[
+                            dependent_section
+                        ][
+                            block_id
+                        ].add(
+                            *rule_block_dependencies_for_progress_value_source[
+                                dependent_section
+                            ][block_id]
+                        )
+            else:
+                self._when_rules_block_dependencies_by_section_for_progress_value_source[
+                    dependent_section
+                ] = rule_block_dependencies_for_progress_value_source[
+                    dependent_section
+                ]
```

Comment:
This could be simplified by using `.items()` to get the values instead of fetching from the dict every time, and by assigning the long property name to a var to make it more readable since this method is about block dependencies you won't lose context. It could be even simplifier with a default dict but this is good enough.
```suggestion
        dependencies = (
            self._when_rules_block_dependencies_by_section_for_progress_value_source
        )

        for (
            dependent_section,
            section_dependencies_by_block,
        ) in rule_block_dependencies_for_progress_value_source.items():
            if dependent_section in dependencies:
                for block_id, section_ids in section_dependencies_by_block.items():
                    if block_id in dependencies[dependent_section]:
                        dependencies[dependent_section][block_id].add(*section_ids)
            else:
                dependencies[
                    dependent_section
                ] = rule_block_dependencies_for_progress_value_source[dependent_section]
```